Readme.MD
# Cake.Apigee

![https://ci.appveyor.com/api/projects/status/3118flw53uycwffg?svg=true](https://ci.appveyor.com/api/projects/status/3118flw53uycwffg?svg=true)

A Cake AddIn that extends Cake with [Apigee](https://apigee.com) management API commands.

[![cakebuild.net](https://img.shields.io/badge/WWW-cakebuild.net-blue.svg)](http://cakebuild.net/)

## Including addin

```
#addin "Cake.Apigee"
```

## Usage
The following shows a simple example of importing a proxy, other aliases are covered
in further sections.

```csharp
#addin "Cake.Apigee"

...

Task("DeployProxy")
    .Does(() =>
{    
    var apigeeCredentials = new Credentials
        {
            Username = "me@me.com",
            Password = "mypassword"
        };

    var result = ImportProxy(
                    "myorg", 
                    "myapi", 
                    File("apiproxy.zip"), 
                    new ImportProxySettings { Credentials = apigeeCredentials });
}
```

### InstallNodePackagedModules

This command effectively runs ```npm install``` on a proxy that contains a NodeJS application (e.g. an a127 project).
It can be useful to do this when the proxy exceeds the default 15Mb limit (you'll need to manually delete the node_modules folder prior
to importing the proxy). Further reading:

* Proxy size limit https://community.apigee.com/questions/10850/api-proxy-size-limitation.html
* Management API function http://docs.apigee.com/management/apis/post/organizations/%7Borg_name%7D/apis/%7Bapi_name%7D/revisions/%7Brevision_num%7D/npm-0

```csharp
InstallNodePackagedModules(
                "myorg", 
                result, 
                new InstallNodePackagedModulesSettings { Credentials = apigeeCredentials });
```

### DeployProxy

This command deploys an already imported proxy to a specified environment.
By default, the DeployProxySettings will use override=true and an allowance of
15 seconds for existing client calls to complete to minimise any downtime during the switchover.

Further reading:
* Seamless deployment http://docs.apigee.com/api-services/content/deploy-api-proxies-using-management-api#seamless

```csharp
DeployProxy(
        "myorg", 
        "dev", 
        result, 
        new DeployProxySettings { Credentials = apigeeCredentials });
```

### GetApiProxy

```csharp
var proxy = GetApiProxy(
        "myorg",             
        "my-proxy", 
        new GetApiProxySettings { Credentials = apigeeCredentials });
```

### DeleteApiProxyRevision

```csharp
DeleteApiProxyRevision(
        "myorg",             
        "my-proxy", 
        "123",
        new DeleteApiProxyRevisionSettings { Credentials = apigeeCredentials });
```

### DeleteAllUndeployedApiProxyRevisions

The underlying code attempts to delete all proxies and Apigee prevents
the deleting of currently deployed proxies.

```csharp
DeleteAllUndeployedApiProxyRevisions(
        "myorg",             
        "my-proxy",             
        new DeleteAllUndeployedApiProxyRevisionsSettings { Credentials = apigeeCredentials });
```

### CreateKeyValueMap

Create a key value map either at the organisation level or environment level (when 
an environment is specified in CreateKeyValueMapSettings).

Important: 

* see Permissions section below on special permissions required to modify Key Value Maps.
* apigee-access > 1.4.0 is required for KVM use from NodeJS https://community.apigee.com/questions/19816/accessing-kvm-from-node-js.html


```csharp
CreateKeyValueMap(
    "myorg",
    new KeyValueMap
    {
        Name = "myMap",
        Encrypted = false,
        Entry = new[]
        {
            new KeyValueMapEntry { Name = "myKey", Value = "myValue" }
        }
    },
    new CreateKeyValueMapSettings { 
        Credentials = apigeeCredentials, 
        Environment = "dev" });
```

### DeleteKeyValueMap

Specify the environment in DeleteKeyValueMapSettings if this is an environment specific key value map.

```csharp
DeleteKeyValueMap(
        "myorg",
        "myMap",
        new DeleteKeyValueMapSettings { 
            Credentials = apigeeCredentials, 
            Environment = "dev" });
```
            
### ListKeyValueMaps

```csharp
IEnumerable<string> mapNames = ListKeyValueMaps(
        "myorg",
        new ListKeyValueMapsSettings { 
            Credentials = apigeeCredentials, 
            Environment = "dev" });
```

# Notes
For more information on the behaviour of these commands, refer to the Apigee management API documentation at <http://docs.apigee.com/api-services/content/api-reference-getting-started>

# Debugging

There is a "Debug" flag on each of the Settings objects that will cause the Apigee response to be dumped to the console.

# Permissions
## Key Value Maps
Apigee requires additional permissions to manage Key Value Maps that are not settable in their UI. These permissions can be added to a role
by using http://docs.apigee.com/management/apis/post/organizations/%7Borg_name%7D/userroles/%7Brole_name%7D/resourcepermissions

Post the following JSON to https://api.enterprise.apigee.com/v1/organizations/{org}/userroles/{yourrole}/resourcepermissions:

```json
{
 "resourcePermission" : [ 
   {
    "path" : "/keyvaluemaps",
    "permissions" : [ "get","put","delete" ]
   }, 
   {
    "path" : "/environments/*/keyvaluemaps",
    "permissions" : [ "get","put","delete" ]
   }
  ]
}
```

# Future
## Removal of node_modules
An option to remove the node_modules from the zip prior to uploading the proxy. This is to mitigate exceeding the 15Mb proxy limit. Currently this must be done using standard Cake Zip functions and the above *InstallNodePackagedModules* run once the proxy has imported.  

[![Follow @cybercohesion](https://img.shields.io/badge/Twitter-Follow%20%40cybercohesion-blue.svg)](https://twitter.com/intent/follow?screen_name=cybercohesion)


ReleaseNotes.MD
## New in 0.15.0 (Released 08/6/2018)
- Appveyor fix

## New in 0.14.0 (Released 08/6/2018)
- Upgrade to Cake 0.28.0
- Upgrade to .netstandard 2.0

## New in 0.13.0 (Released 19/1/2018)
- Set TLS 1.2 to fix exception of getting:
  "An error occurred while sending the request." with inner error of:
  "Authentication failed because the remote party has closed the transport stream."

## New in 0.12.0 (Released 8/11/2017)
- default the ImportProxy to validate the proxy. It can be set back to previous behaviour of non-validation using
  setting the ImportProxySettings.Validate property to false.

## New in 0.11.0 (Released 28/9/2017)
- no functional change, updated to using Cake v0.22.2.

## New in 0.10.0 (Released 1/8/2017)
- no functional change, using the cake.contrib icon.

## New in 0.9.0 (Released 25/4/2017)
- increased the HTTP client timeout to 3 minutes to allow long running commands such as 'npm install' to complete.

## New in 0.8.0 (Released 24/4/2017)
- updated the XML docs and also output http status if a proxy deploy fails to assist in diagnosis.

## New in 0.7.0 (Released 16/3/2017)
- fixed bug where deploy result failed to deserialize if no other revisions existed.

## New in 0.6.0 (Released 16/3/2017)
- test release.

## New in 0.5.0 (Released 01/03/2017)
- added CreateKeyValueMap, ListKeyValueMaps and DeleteKeyValueMap which support
  organisation and environment scope.

## New in 0.4.0 (Released 19/02/2017)
- added DeleteAllUndeployedApiProxyRevisions, GetApiProxy and DeleteApiProxyRevision
- improved error messages to include the URL and method of any failed calls

## New in 0.3.0 (Released 19/02/2017)
- improved error reporting
- return and list the packages restored by npm in Apigee

## New in 0.2.0 (Released 17/02/2017)
- Fix for nuget package folder to lib\net45

## New in 0.1.0 (Released 16/02/2017)
- Added Cake Aliases Apigee Methods for ImportProxy, DeployProxy and InstallNodePackagedModules
- Setup build via AppVeyor for automatic deployment to NuGet
- Limited Unit and Integration Tests added for code coverage

src/Cake.Apigee/ApigeeAliases.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Net.Http;
using Cake.Apigee.Contracts;
using Cake.Apigee.Services;
using Cake.Core;
using Cake.Core.Annotations;
using Cake.Core.Diagnostics;
using Cake.Core.IO;

namespace Cake.Apigee
{
    /// <summary>
    /// Contains functionality for proxy import and deployments within Apigee.
    /// </summary>
    [CakeAliasCategory("Apigee")]
    public static class ApigeeAliases
    {
        public static ApigeeProxyManagementService ApigeeProxyManagementService { get; set; } = new ApigeeProxyManagementService(new HttpClient());

        /// <summary>
        /// Import a proxy into Apigee from a zip file. The zip file should contain an apiproxy folder at the top level.
        /// </summary>
        /// <example>
        /// <para>Cake task:</para>
        /// <code>
        /// <![CDATA[
        /// var orgName = "myorg";
        /// var apigeeCredentials = new Credentials
        /// {
        ///    Username = "me@me.com",
        ///    Password = "mypassword"
        /// };
        ///
        /// var result = ImportProxy(
        ///                orgName,
        ///                "myapi",
        ///                File("apiproxy.zip"),
        ///                new ImportProxySettings { Credentials = apigeeCredentials });
        /// ]]>
        /// </code>
        /// </example>
        /// <remarks>
        /// This calls the following API:
        /// https://docs.apigee.com/management/apis/post/organizations/%7Borg_name%7D/apis-0
        /// </remarks>
        /// <param name="ctx">The Cake context.</param>
        /// <param name="orgName">Your org name as shown in Apigee.</param>
        /// <param name="proxyName">Name of the proxy.</param>
        /// <param name="proxyZipfile">Path to the zip.</param>
        /// <param name="settings">The settings such as authentication credentials.</param>
        /// <returns>The result of the import such as the revision created.</returns>
        [CakeMethodAlias]
        public static ImportProxyResult ImportProxy(this ICakeContext ctx, string orgName, string proxyName, FilePath proxyZipfile, ImportProxySettings settings = null)
        {
            ctx.Log.Information("Import proxy " + proxyZipfile);

            return Run(() => ApigeeProxyManagementService.ImportProxyAsync(ctx, orgName, proxyName, proxyZipfile, settings ?? new ImportProxySettings()).Result);
        }

        /// <summary>
        /// Deploy a proxy within Apigee to an environment in Apigee using the results of an import. By default, if settings
        /// are supplied, they will use override and a 15 second delay to minimize downtime as described in:
        /// http://docs.apigee.com/api-services/content/deploy-api-proxies-using-management-api#seamless
        /// </summary>
        /// <example>
        /// <para>Cake task:</para>
        /// <code>
        /// <![CDATA[
        /// var orgName = "myorg";
        /// var apigeeCredentials = new Credentials
        /// {
        ///    Username = "me@me.com",
        ///    Password = "mypassword"
        /// };
        ///
        /// DeployProxy(
        ///     orgName,
        ///     "dev",
        ///     result,
        ///     new DeployProxySettings { Credentials = apigeeCredentials });
        /// ]]>
        /// </code>
        /// </example>
        /// <remarks>
        /// This calls the following API:
        /// https://docs.apigee.com/management/apis/post/organizations/%7Borg_name%7D/environments/%7Benv_name%7D/apis/%7Bapi_name%7D/revisions/%7Brevision_number%7D/deployments
        /// </remarks>
        /// <param name="ctx">The Cake context.</param>
        /// <param name="orgName">Your org name as shown in Apigee.</param>
        /// <param name="envName">The name of the environment as shown in Apigee.</param>
        /// <param name="importResult">Using the result of the import saves re-specifying the name and revision.</param>
        /// <param name="settings">The setting such as authentication credentials.</param>
        /// <returns></returns>
        [CakeMethodAlias]
        public static DeployProxyResult DeployProxy(this ICakeContext ctx, string orgName, string envName, ImportProxyResult importResult, DeployProxySettings settings = null)
        {
            ctx.Log.Information("Deploy api " + importResult.Name);

            return DeployProxy(
                ctx,
                orgName,
                envName,
                importResult.Name,
                importResult.Revision,
                settings);
        }

        /// <summary>
        /// Deploy a proxy to Apigee using a specific revision number.
        /// </summary>
        /// <example>
        /// <para>Cake task to deploy revision 123 of myapi to the dev environment for myorg:</para>
        /// <code>
        /// <![CDATA[
        /// var orgName = "myorg";
        /// var apigeeCredentials = new Credentials
        /// {
        ///    Username = "me@me.com",
        ///    Password = "mypassword"
        /// };
        ///
        /// DeployProxy(
        ///     orgName,
        ///     "dev",
        ///     "myapi",
        ///     "123",
        ///     new DeployProxySettings { Credentials = apigeeCredentials });
        /// ]]>
        /// </code>
        /// </example>
        /// <remarks>
        /// This calls the following API:
        /// https://docs.apigee.com/management/apis/post/organizations/%7Borg_name%7D/environments/%7Benv_name%7D/apis/%7Bapi_name%7D/revisions/%7Brevision_number%7D/deployments
        /// </remarks>
        /// <param name="ctx">The Cake context.</param>
        /// <param name="orgName">Your org name as shown in Apigee.</param>
        /// <param name="envName">The name of the environment as shown in Apigee.</param>
        /// <param name="apiName">Name of the proxy/api.</param>
        /// <param name="revisionNumber">The revision number of the proxy.</param>
        /// <param name="settings">The deployment settings such as the credentials. At least pass new settings to get the seamless deployment parameters described above.</param>
        /// <returns>Result of the deployment.</returns>
        [CakeMethodAlias]
        public static DeployProxyResult DeployProxy(
            this ICakeContext ctx,
            string orgName,
            string envName,
            string apiName,
            string revisionNumber,
            DeployProxySettings settings = null)
        {
            return Run(() => ApigeeProxyManagementService.DeployProxyAsync(ctx, orgName, envName, apiName, revisionNumber, settings).Result);
        }

        /// <summary>
        /// Run "npm install" on NodeJS embedded as a resource in the proxy. This can be useful if your proxy exceeds the 15MB limit imposed by Apigee.
        /// </summary>
        /// <example>
        /// <para>Cake task:</para>
        /// <code>
        /// <![CDATA[
        /// var orgName = "myorg";
        /// var apigeeCredentials = new Credentials
        /// {
        ///    Username = "me@me.com",
        ///    Password = "mypassword"
        /// };
        ///
        /// InstallNodePackagedModules(
        ///                    orgName,
        ///                    "myapi",
        ///                    "123",
        ///                    new InstallNodePackagedModulesSettings { Credentials = apigeeCredentials
        /// });        
        /// ]]>
        /// </code>
        /// </example>
        /// <remarks>
        /// This runs the following API with the command "install":
        /// https://docs.apigee.com/management/apis/post/organizations/%7Borg_name%7D/apis/%7Bapi_name%7D/revisions/%7Brevision_num%7D/npm-0
        /// </remarks>
        /// <param name="ctx">The Cake context.</param>
        /// <param name="orgName">Your org name as shown in Apigee.</param>
        /// <param name="proxyName">Name of the proxy.</param>
        /// <param name="revisionNumber">The revision to deploy.</param>
        /// <param name="settings">The settings such as the authentication credentials.</param>
        /// <returns>The list of installed modules.</returns>
        [CakeMethodAlias]
        public static NodePackagedModuleMetadata[] InstallNodePackagedModules(
            this ICakeContext ctx,
            string orgName,
            string proxyName,
            string revisionNumber,
            InstallNodePackagedModulesSettings settings = null)
        {
            return ApigeeProxyManagementService.InstallNodePackagedModules(
                ctx,
                orgName,
                proxyName,
                revisionNumber,
                settings).Result;
        }

        /// <summary>
        /// Run "npm install" on NodeJS embedded as a resource in the proxy based on a previous import. This can be useful if your proxy exceeds the 15MB limit imposed by Apigee.
        /// </summary>
        /// <example>
        /// <para>Cake task:</para>
        /// <code>
        /// <![CDATA[
        /// var orgName = "myorg";
        /// var apigeeCredentials = new Credentials
        /// {
        ///    Username = "me@me.com",
        ///    Password = "mypassword"
        /// };
        ///
        /// InstallNodePackagedModules(
        ///                    orgName,
        ///                    importResult,
        ///                    new InstallNodePackagedModulesSettings { Credentials = apigeeCredentials
        /// });        
        /// ]]>
        /// </code>
        /// </example>        
        /// <remarks>
        /// This runs the following API with the command "install":
        /// https://docs.apigee.com/management/apis/post/organizations/%7Borg_name%7D/apis/%7Bapi_name%7D/revisions/%7Brevision_num%7D/npm-0
        /// </remarks>
        /// <param name="ctx">The Cake context.</param>
        /// <param name="orgName">Your org name as shown in Apigee.</param>
        /// <param name="importResult">The result of an import.</param>
        /// <param name="settings">Settings such as the authentication credentials.</param>
        /// <returns>The list of modules and versions that were installed.</returns>
        [CakeMethodAlias]
        public static NodePackagedModuleMetadata[] InstallNodePackagedModules(
            this ICakeContext ctx,
            string orgName,
            ImportProxyResult importResult,
            InstallNodePackagedModulesSettings settings = null)
        {
            return Run(() => ApigeeProxyManagementService.InstallNodePackagedModules(
                ctx,
                orgName,
                importResult.Name,
                importResult.Revision,
                settings).Result);
        }

        /// <summary>
        /// Retrieve API proxy metadata such as the list of revisions and when the proxy was created.
        /// </summary>
        /// <example>
        /// <para>Cake task:</para>
        /// <code>
        /// <![CDATA[
        /// var orgName = "myorg";
        /// var apigeeCredentials = new Credentials
        /// {
        ///    Username = "me@me.com",
        ///    Password = "mypassword"
        /// };
        ///
        /// var proxyMetadata = GetApiProxy(
        ///                    orgName,
        ///                    "myapi",
        ///                    new GetApiProxySettings { Credentials = apigeeCredentials });
        /// ]]>
        /// </code>
        /// </example>        
        /// <param name="ctx">The Cake context.</param>
        /// <param name="orgName">Your org name as shown in Apigee.</param>
        /// <param name="proxyName">Name of the proxy.</param>
        /// <param name="settings">The settings such as authentication credentials.</param>
        /// <returns>The proxy metadata.</returns>
        [CakeMethodAlias]
        public static ApiProxy GetApiProxy(
           this ICakeContext ctx,
           string orgName,
           string proxyName,
           GetApiProxySettings settings = null)
        {
            return Run(() => ApigeeProxyManagementService.GetApiProxy(
                ctx,
                orgName,
                proxyName,                
                settings).Result);
        }

        /// <summary>
        /// Delete a particular revision of a proxy.
        /// </summary>
        /// <example>
        /// <para>Cake task:</para>
        /// <code>
        /// <![CDATA[
        /// var orgName = "myorg";
        /// var apigeeCredentials = new Credentials
        /// {
        ///    Username = "me@me.com",
        ///    Password = "mypassword"
        /// };
        ///
        /// DeleteApiProxyRevision(
        ///                    orgName,
        ///                    "myapi",
        ///                    "123",
        ///                    new DeleteApiProxyRevisionSettings { Credentials = apigeeCredentials });
        /// ]]>
        /// </code>
        /// </example>        
        /// <param name="ctx">The Cake context.</param>
        /// <param name="orgName">Your org name as shown in Apigee.</param>
        /// <param name="proxyName">The name of the proxy.</param>
        /// <param name="revisionNumber">The revision to delete.</param>
        /// <param name="settings">Settings such as the authentication credentials.</param>
        [CakeMethodAlias]
        public static void DeleteApiProxyRevision(
           this ICakeContext ctx,
           string orgName,
           string proxyName,
           string revisionNumber,
           DeleteApiProxyRevisionSettings settings = null)
        {
            Run(() => ApigeeProxyManagementService.DeleteApiProxyRevision(
                ctx,
                orgName,
                proxyName,
                revisionNumber,
                settings).Result);
        }

        /// <summary>
        /// Delete all undeployed proxy revisions. This attempts to delete ALL proxy revisions but Apigee won't let
        /// already deployed revisions be removed.
        /// </summary>
        /// <example>
        /// <para>Cake task:</para>
        /// <code>
        /// <![CDATA[
        /// var orgName = "myorg";
        /// var apigeeCredentials = new Credentials
        /// {
        ///    Username = "me@me.com",
        ///    Password = "mypassword"
        /// };
        ///
        /// DeleteAllUndeployedApiProxyRevisions(
        ///                    orgName,
        ///                    "myapi",
        ///                    new DeleteAllUndeployedApiProxyRevisionsSettings { Credentials = apigeeCredentials });
        /// ]]>
        /// </code>
        /// </example>        
        /// <remarks>
        /// This makes calls to the following API for every revision of the proxy:
        /// https://docs.apigee.com/management/apis/delete/organizations/%7Borg_name%7D/apis/%7Bapi_name%7D/revisions/%7Brevision_number%7D
        /// </remarks>
        /// <param name="ctx">The Cake context.</param>
        /// <param name="orgName">Your org name as shown in Apigee.</param>
        /// <param name="proxyName">The name of the proxy.</param>
        /// <param name="settings">The settings such as the authentication credentials.</param>
        [CakeMethodAlias]
        public static void DeleteAllUndeployedApiProxyRevisions(
           this ICakeContext ctx,
           string orgName,
           string proxyName,
           DeleteAllUndeployedApiProxyRevisionsSettings settings = null)
        {
            Run(() => ApigeeProxyManagementService.DeleteAllUndeployedApiProxyRevisions(
                ctx,
                orgName,
                proxyName,                
                settings).Wait());
        }

        /// <summary>
        /// Create an organisation or environment scoped key value map (KVM) in Apigee. 
        /// </summary>
        /// <example>
        /// <para>Cake task:</para>
        /// <code>
        /// <![CDATA[
        /// var orgName = "myorg";
        /// var apigeeCredentials = new Credentials
        /// {
        ///    Username = "me@me.com",
        ///    Password = "mypassword"
        /// };
        ///
        /// CreateKeyValueMap(
        ///                 orgName,
        ///                 "myapi",
        ///                 new KeyValueMap
        ///                 {
        ///                     Name = "myMap",
        ///                     Encrypted = false,
        ///                     Entry = new[]
        ///                     {
        ///                         new KeyValueMapEntry { Name = "myKey", Value = "myValue" }
        ///                     }
        ///                  },
        ///                  new CreateKeyValueMapSettings { Credentials = apigeeCredentials, Environment = "dev" });
        /// ]]>
        /// </code>
        /// </example>        
        /// <param name="ctx">The Cake context.</param>
        /// <param name="orgName">Your org name as shown in Apigee.</param>
        /// <param name="keyValueMap">The values to set in the key value map.</param>
        /// <param name="settings">Settings such as credentials and an environment name if environment scoped.</param>
        [CakeMethodAlias]
        public static void CreateKeyValueMap(
           this ICakeContext ctx,
           string orgName,
           KeyValueMap keyValueMap,
           CreateKeyValueMapSettings settings = null)
        {
            Run(() => ApigeeProxyManagementService.CreateKeyValueMap(
                ctx,
                orgName,
                keyValueMap,
                settings).Wait());
        }

        /// <summary>
        /// Delete a key value map (KVM).
        /// </summary>
        /// <example>
        /// <para>Cake task:</para>
        /// <code>
        /// <![CDATA[
        /// var orgName = "myorg";
        /// var apigeeCredentials = new Credentials
        /// {
        ///    Username = "me@me.com",
        ///    Password = "mypassword"
        /// };
        ///
        /// DeleteKeyValueMap(
        ///                    orgName,
        ///                    "myMap",
        ///                    new DeleteKeyValueMapSettings { Credentials = apigeeCredentials, Environment = "dev" });
        /// ]]>
        /// </code>
        /// </example>        
        /// <param name="ctx">The Cake context.</param>
        /// <param name="orgName">Your org name as shown in Apigee.</param>
        /// <param name="keyValueMapName">The name of the Key Value Map to delete.</param>
        /// <param name="settings">The settings such as authentication credentials and environment for environment scoped maps.</param>
        [CakeMethodAlias]
        public static void DeleteKeyValueMap(
           this ICakeContext ctx,
           string orgName,
           string keyValueMapName,
           DeleteKeyValueMapSettings settings = null)
        {
            Run(() => ApigeeProxyManagementService.DeleteKeyValueMap(
                ctx,
                orgName,
                keyValueMapName,
                settings).Wait());
        }

        /// <summary>
        /// List all of the key value maps at either organisation or environment scope.
        /// </summary>
        /// <example>
        /// <para>Cake task:</para>
        /// <code>
        /// <![CDATA[
        /// var orgName = "myorg";
        /// var apigeeCredentials = new Credentials
        /// {
        ///    Username = "me@me.com",
        ///    Password = "mypassword"
        /// };
        ///
        /// IEnumerable<string> list = ListKeyValueMaps(
        ///                    orgName,
        ///                    new ListKeyValueMapsSettings { Credentials = apigeeCredentials, Environment = "dev" });
        /// ]]>
        /// </code>
        /// </example>        
        /// <param name="ctx">The Cake context.</param>
        /// <param name="orgName">Your org name as shown in Apigee.</param>
        /// <param name="settings">Settings such as the Apigee authentication credentials.</param>
        /// <returns>The list of key value map names.</returns>
        [CakeMethodAlias]
        public static IEnumerable<string> ListKeyValueMaps(
           this ICakeContext ctx,
           string orgName,           
           ListKeyValueMapsSettings settings = null)
        {
            return Run(() => ApigeeProxyManagementService.ListKeyValueMaps(
                ctx,
                orgName,                
                settings).Result);
        }

        private static void Run(Action function)
        {
            try
            {
                function.Invoke();
            }
            catch (AggregateException aggEx)
            {
                throw new Exception(string.Join("; ", aggEx.Flatten().InnerExceptions.Select(ex => ex.Message)));
            }
        }

        private static T Run<T>(Func<T> function)
        {
            try
            {
                return function.Invoke();
            }
            catch (AggregateException aggEx)
            {                
                throw new Exception(string.Join("; ", aggEx.Flatten().InnerExceptions.Select(ex => ex.Message)));
            }
        }
    }
}
